sycl/doc/extensions/proposed/sycl_ext_intel_fpga_datapath.asciidoc

```diff
@@ -0,0 +1,239 @@
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Representation of FPGA memory
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+template <typename T>
+class fpga_datapath {
+  ...
+----
+
+`fpga_datapath` is a class template, parameterized by the type of the underlying
+allocation _T_ which encodes the size of the allocation.
+
+`fpga_datapath` initializes the underlying object of type `T` with the values passed into its constructor.
+
+[NOTE]
+====
+
+`fpga_datapath` has similar semantics to a reference wrapper. An fpga_datapath templated on a type T can implicitly convert to T, thereby allowing it to be used in most places that an object of type T could be used. The dot operator
+(`operator.`) cannot be overloaded, so a `get()` member is provided to allow a
+reference to be extracted directly when needed.
+====
+
+=== FPGA Datapath Interface
+
+The section below and the table following describe the constructors and member functions for `fpga_datapath`.
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+
+  template <typename T>
+  class fpga_datapath {
+  public:
+    template <typename... S> constexpr fpga_datapath(S... args);
+    
+    // Access the underlying data
+    operator T &() noexcept;
+    operator const T &() const noexcept;
+
+    T &get() noexcept;
+    const T &get() const noexcept;
+
+    // Enable assignments from underlying type
+    fpga_datapath &operator=(const T &) noexcept;
+
+    // Note that there is no need for "fpga_datapath" to define member functions
+    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
+    // "T" need only define these operators as non-member functions. Because
+    // there is an implicit conversion from "fpga_datapath" to "T&", the operations
+    // can be applied to objects of type "fpga_datapath<T>"
+  };
+
+} // namespace sycl::ext::intel::experimental
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename ... S>
+constexpr fpga_datapath(S ... args);
+----
+|
+Constructs a `fpga_datapath` object, and implicit storage for `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T&() noexcept;
+operator const T&() const noexcept;
+----
+|
+Implicit conversion to a reference to the underlying `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+T& get() noexcept;
+const T& get() const noexcept;
+----
+|
+Returns a reference to the underlying `T`.
+
+|===
+
+=== Restrictions on creating fpga_datapath objects
+
+There are restrictions on how the application can create objects of type
+`fpga_datapath`. Applications that violate these restrictions are ill-formed.
+
+* The `fpga_datapath` variable can only have accesses from a single kernel. This can happen when `device_global` is templated on a `fpga_datapath` type. 
```

Comment:
I don't think we need this restriction in the spec.  We've already said that the class is a "request", implying that an implementation can ignore it in some cases.  We could document in our docs that this is one case in which our implementation ignores it.  I worry that with this wording ("... are ill-formed"), the frontend folks will interpret it as their responsibility to error out on any program that accesses such a variable in multiple kernels which is not typically a task you'd want to perform in the FE.

Reply:
Note that if you remove this, you should also remove the reference to the device_global extension in the dependencies section because this is the only mention of device_global in the spec.

Reply:
I decided to call it out explicitly. I think that is clearer for understanding.

Reply:
After thinking about it more, I don't think this is implementable as written.  If the code is ill-formed (without the "no diagnostic required" qualifier) then the FE must diagnose it.  But this situation is impossible for the FE to diagnose without building a call graph of the kernel (which the FE team has already told us they don't think should be the responsibility of the FE) and is completely impossible to do at compile time at all for any target that supports the function pointers extension.  

Furthermore, we don't need to say this in the spec and it adds significant complexity to the spec by creating a link to device_globals.  We've already said that the class is a hint and so we are within our rights to not follow it in some cases.  We should instead mention this case in our docs and emit a warning when the user hits it.  This also future-proofs the spec against possible improvements to the compiler that would allow fpga_datapath variables to be accessed by more than one kernel (for example if we supported kernel fusion).  In such a case we'd merely have to update or docs and implementation rather than the spec itself.

Reply:
removed.